### DISCUSSION TITLE
[Question]: LibreChat not working behind nginx reverse proxy with basic_auth

### DISCUSSION THREAD

**Discussion by jerkstorecaller** [danny-avila/LibreChat#890] [Troubleshooting] [answered]
### What is your question?

I'm pretty inexperienced with web stuff so bear with me.

I'm trying to self-host Librechat the way I do my other apps: behind an nginx webserver which acts as a reverse proxy. I put everything at subdomain.domain.com, in this case I wanted gpt.domain.com. nginx uses basic_auth to protect access to each subdomain.

Librechat loads, but the API calls are not working, so it does nothing. If I access Librechat directly (on plain HTTP port 3080 instead of going through nginx), it works fine. But I don't want to do that for security reasons.

My nginx.con uses the exact same server block for Librechat as for my other stuff:

Nginx's access.log shows that nearly all calls after the initial page load, such as 'GET /api/whatever', are returning 401 (which I think should show an error btw). I modified nginx's log template to add info (user=, auth=, returncode=), and I notice that the failed calls all have a different auth style. On my other apps (eg QBittorrent), I always see user=myusername, auth=Basic, and all calls 200 OK. LibreChat is different, it's an empty user=-, and auth=Bearer instead of Basic with a weird token that's like 5x longer.

Can I do anything about this? Is Librechat compatible with basic_auth?

relevant nginx.conf block:
```
#https://gpt.domain.com
server {
	listen 443 ssl;
	server_name gpt.*;
	include /config/nginx/ssl.conf;

	location / {
		auth_basic "Restricted";
		auth_basic_user_file /config/nginx/.htpasswd;
		include /config/nginx/proxy_params.conf;
		proxy_pass http://LibreChat:3080/;
	}
}
```

access.log
```
#Successful calls when loading
101.102.103.104 [06/Sep/2023:18:14:10 +0000] user=myusername server=gpt.* auth=Basic XaGakH5n19asFkji3bnjaRajkhsajhj"GET / HTTP/1.1" returncode=304 ref="-" 
101.102.103.104 [06/Sep/2023:18:14:11 +0000] user=myusername server=gpt.* auth=Basic XaGakH5n19asFkji3bnjaRajkhsajhj"GET /assets/vendor-647c35c8.js HTTP/1.1" returncode=304 ref="https://gpt.domain.com/" 
101.102.103.104 [06/Sep/2023:18:14:11 +0000] user=myusername server=gpt.* auth=Basic XaGakH5n19asFkji3bnjaRajkhsajhj"GET /assets/index-fbf91cb2.js HTTP/1.1" returncode=304 ref="https://gpt.domain.com/" 
101.102.103.104 [06/Sep/2023:18:14:11 +0000] user=myusername server=gpt.* auth=Basic XaGakH5n19asFkji3bnjaRajkhsajhj"GET /assets/index-f96e0051.css HTTP/1.1" returncode=304 ref="https://gpt.domain.com/" 

#401 calls after page load, except for, weirdly enough, a single proper basic_auth call to /api/endpoints
101.102.103.104 [06/Sep/2023:18:14:11 +0000] user=- server=gpt.* auth=Bearer LOOOOOOOOONG-AUTH-TOKEN "GET /api/convos?pageNumber=1 HTTP/1.1" returncode=401 ref="https://gpt.domain.com/" 
101.102.103.104 [06/Sep/2023:18:14:11 +0000] user=myusername server=gpt.* auth=Basic XaGakH5n19asFkji3bnjaRajkhsajhj"GET /api/search/enable HTTP/1.1" returncode=200 ref="https://gpt.domain.com/" 
101.102.103.104 [06/Sep/2023:18:14:11 +0000] user=- server=gpt.* auth=Bearer LOOOOOOOOONG-AUTH-TOKEN "GET /api/user HTTP/1.1" returncode=401 ref="https://gpt.domain.com/chat/new" 
101.102.103.104 [06/Sep/2023:18:14:11 +0000] user=- server=gpt.* auth=Bearer LOOOOOOOOONG-AUTH-TOKEN "GET /api/config HTTP/1.1" returncode=401 ref="https://gpt.domain.com/chat/new" 
101.102.103.104 [06/Sep/2023:18:14:12 +0000] user=- server=gpt.* auth=Bearer LOOOOOOOOONG-AUTH-TOKEN "GET /api/search/enable HTTP/1.1" returncode=401 ref="https://gpt.domain.com/chat/new" 
101.102.103.104 [06/Sep/2023:18:14:12 +0000] user=- server=gpt.* auth=Bearer LOOOOOOOOONG-AUTH-TOKEN "GET /api/endpoints HTTP/1.1" returncode=401 ref="https://gpt.domain.com/chat/new" 
101.102.103.104 [06/Sep/2023:18:14:12 +0000] user=myusername server=gpt.* auth=Basic XaGakH5n19asFkji3bnjaRajkhsajhj"GET /api/endpoints HTTP/1.1" returncode=200 ref="https://gpt.domain.com/" 
101.102.103.104 [06/Sep/2023:18:14:13 +0000] user=- server=gpt.* auth=Bearer LOOOOOOOOONG-AUTH-TOKEN "GET /api/config HTTP/1.1" returncode=401 ref="https://gpt.domain.com/login" 
101.102.103.104 [06/Sep/2023:18:14:13 +0000] user=- server=gpt.* auth=Bearer LOOOOOOOOONG-AUTH-TOKEN "GET /api/convos?pageNumber=1 HTTP/1.1" returncode=401 ref="https://gpt.domain.com/chat/new" 
...
```

### More Details

None

### What is the main subject of your question?

User System/OAuth

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
LibreChat has its own built-in authentication system, so it won't know how to handle the one you're passing through to the API.

I found this ChatGPT answer helpful here:

---
The issue you're facing is due to a conflict between the `Basic` authentication you've set up in Nginx and the `Bearer` token authentication that LibreChat uses for its API calls.

Here's a breakdown:

1. **Basic Authentication**: This is the authentication method you've set up in Nginx. When you access the site, your browser prompts you for a username and password. Once entered, the browser sends these credentials with every request to the server, encoded in the `Authorization` header as `Basic <base64-encoded-credentials>`.

2. **Bearer Token Authentication**: This is a common authentication method for APIs. After an initial login, the server provides a token (often a JWT or similar) that the client then sends with every subsequent request to prove its identity. This token is sent in the `Authorization` header as `Bearer <token>`.

The problem arises when LibreChat tries to make an API call. It sends the `Bearer` token in the `Authorization` header, but Nginx is expecting the `Basic` credentials. Since the `Bearer` token doesn't match the `Basic` credentials, Nginx returns a `401 Unauthorized` response.

Here's how you can resolve the issue:

1. **Exclude API Calls from Basic Authentication**: You can modify your Nginx configuration to exclude the `/api/` path from the `Basic` authentication. This way, only the `Bearer` token will be checked for API calls.

```nginx
server {
    listen 443 ssl;
    server_name gpt.*;
    include /config/nginx/ssl.conf;

    location / {
        auth_basic "Restricted";
        auth_basic_user_file /config/nginx/.htpasswd;
        include /config/nginx/proxy_params.conf;
        proxy_pass http://LibreChat:3080/;
    }

    location ~ ^/api/ {
        auth_basic off;
        include /config/nginx/proxy_params.conf;
        proxy_pass http://LibreChat:3080/;
    }
}
```

2. **Ensure LibreChat Handles Authentication**: By excluding the `/api/` path from Nginx's `Basic` authentication, you're relying on LibreChat to handle its own authentication for API calls. Ensure that LibreChat is set up to require and validate the `Bearer` token for all API calls.

> Danny: it does.

4. **Consider Additional Security Measures**: If you're concerned about security, consider other measures such as using HTTPS for LibreChat, setting up a Web Application Firewall (WAF), or using other authentication mechanisms that are compatible with both Nginx and LibreChat.

After making these changes, restart Nginx and test again. The API calls from LibreChat should now work correctly through the reverse proxy.

**Comment by danny-avila** [comment]
One thing to note, when using NGINX, it's expected to proxy pass to API still, so a working file may look more like this, again according to GPT.

---

Given the default Nginx configuration for LibreChat, it's clear that the application expects to be directly proxied without any additional authentication layers. The default configuration is set up to forward all `/api` requests directly to the application's API endpoint and all other requests to the application's main endpoint.

To integrate Basic Authentication while ensuring that LibreChat's API calls work correctly, you can modify the configuration as follows:

1. **Add SSL Configuration**: If you're planning to use HTTPS (which is recommended), ensure you have the SSL configuration lines uncommented and correctly pointing to your SSL certificate and key.

2. **Integrate Basic Authentication for Non-API Calls**: You can set up Basic Authentication for all non-API paths. This way, when you access the main page, you'll be prompted for a username and password, but the internal API calls made by LibreChat will bypass this authentication.

Here's a modified configuration based on your requirements:

```nginx
server {
    listen 80;
    listen 443 ssl;

#    ssl_certificate /etc/nginx/ssl/nginx.crt;
#    ssl_certificate_key /etc/nginx/ssl/nginx.key;

    server_name gpt.domain.com;  # Change this to your domain

    # Exclude API calls from Basic Authentication
    location ~ ^/api/ {
        auth_basic off;
        proxy_pass http://api:3080/api;
        include /config/nginx/proxy_params.conf;
    }

    # All other requests will have Basic Authentication
    location / {
        auth_basic "Restricted";
        auth_basic_user_file /config/nginx/.htpasswd;
        proxy_pass http://api:3080;
        include /config/nginx/proxy_params.conf;
    }
}
```

A few things to note:

- Ensure that the paths to your SSL certificate and key are correct.
- The `server_name` directive should be set to your domain.
- The `include /config/nginx/proxy_params.conf;` line is assumed to contain common proxy settings. If you don't have such a file, you might need to add proxy settings directly, such as `proxy_set_header Host $host;` and other related directives.
- Ensure that the path to your `.htpasswd` file is correct.

After making these changes, restart Nginx and test the setup. You should be prompted for Basic Authentication when accessing the main page, but the internal API calls made by LibreChat should work without any issues.

**Comment by jerkstorecaller** [comment, marked as answer]
The reason I use basic_auth is because I have fail2ban installed on my server and it reads nginx's access logs and auto-ban any IP that fails too many nginx logins (HTTP 401). (for example those 401 in my access log would have gotten my own IP banned if I kept using them). 

Instead of worrying whether any of the many apps I self-host is properly secured by the developer, I just let nginx+fail2ban take care of auth, and I don't have to care. Even if your app has zero security, I'm unhackable this way. It's a very common setup among casual self-hosters like me who want to set something up and forget about it for months.

I went with option 1. What ChatGPT gave you has a small issue, it gives the error: *"proxy_pass" cannot have URI part in location given by regular expression, or inside named location, or inside "if" statement, or inside "limit_except"*

However, removing the trailing slash at the end of proxy_pass (after :3080) fixes it. So here's the corrected configuration block:

```
#https://gpt.domain.com
server {
	listen 443 ssl;
	listen [::]:443 ssl;
	server_name gpt.*;
	include /config/nginx/ssl.conf;

	location / {
		auth_basic "Restricted";
		auth_basic_user_file /config/nginx/.htpasswd;
		include /config/nginx/proxy_params.conf;
		proxy_pass http://LibreChat:3080;
	}

        #Do not use basic_auth for API calls, since they will use LibreChat's own auth system
	location ~ ^/api/ {
	  auth_basic off;
	  include /config/nginx/proxy_params.conf;
	  proxy_pass http://LibreChat:3080;
	}
}
```

And with this, I'm a happy camper. Thanks Danny! 

I think this solution should be in the docs somewhere ("Using Librechat behind a secured reverse proxy").  These instructions are specific to nginx but Apache and Caddyserver will have the same solution and should be easy to translate for other users. What matters is the solution.

**Reply by danny-avila** [reply]
That’s awesome, glad you figured it out!! Would you mind doing a write up for the docs? If not I can get around to it, but that is a neat use case

**Reply by danny-avila** [reply]
I will have to add that you should also add a rule to add basic auth for the API registration route or just simply set or even omit the following env variable in your .env file: ALLOW_REGISTRATION=false

you can test this with a postman request that registration will not be allowed and therefore none of the api routes will be accessible

**Reply by danny-avila** [reply]
a long time coming but I've gotten the docs updated now: https://github.com/danny-avila/LibreChat/pull/1118

**Reply by passaca** [reply]
May I ask how basic auth will help for "worrying whether any of the many apps I self-host is properly secured" if the API itself is still not protected? Any vulnerability there would still be a risk, or am I missing a crucial point here?

PS: The reason for my comment is that I'm also intending to use LibreChat on a public server and I'm a bit hesitant regarding the security implications.

**Reply by danny-avila** [reply]
@passaca LibreChat is secure against unauthorized requests. I've been running my instance on a fairly public domain with cloudflare protection and using HTTPS, no issues. Granted, there is always more that can be done as you imply.

The added security that Basic Auth offers is to prevent a user from ever becoming authorized in the first place, unless they have those basic auth credentials.

On top of this, to really protect your instance, you need to make use of the automated moderation settings: https://docs.librechat.ai/features/mod_system.html

**Comment by jerkstorecaller** [comment]
No problem. Here's the write-up, copy-paste it where appropriate in the doc:


### Using LibreChat behind a reverse proxy with Basic Authentication

If LibreChat is behind a webserver acting as a reverse proxy with Basic Auth (a common scenario for casual users), LibreChat will not function properly without some extra configuration. You will connect to LibreChat, be prompted to enter Basic Auth credentials, enter your username/password, LibreChat will load, but then you will not get a response from the AI services.

The reason is that LibreChat uses Bearer authentication when calling the backend API at domain.com/api. Because those calls will use Bearer rather than Basic auth, your webserver will view this as unauthenticated connection attempt and return 401.

The solution is to enable Basic Auth, but disable it specifically for the /api/ endpoint. (it's safe because the API calls still require an authenticated user)

You will therefore need to create a new rule that disables Basic Auth for /api/. This rule must be higher priority than the rule activating Basic Auth. 

For example, for nginx, you might do:

```
#https://librechat.domain.com
server {
	listen 443 ssl;
	listen [::]:443 ssl;
	server_name librechat.*;
	include /config/nginx/ssl.conf;

	#all connections to librechat.domain.com require basic_auth
	location / {
	  auth_basic "Access Restricted";
	  auth_basic_user_file /config/nginx/.htpasswd;
	  include /config/nginx/proxy_params.conf;
	  proxy_pass http://127.0.0.1:3080;
	}

    #...except for /api/, which will use LibreChat's own auth system
	location ~ ^/api/ {
	  auth_basic off;
	  include /config/nginx/proxy_params.conf;
	  proxy_pass http://127.0.0.1:3080;
	}
}
```

**Comment by mitya-shepelev** [comment]
Thank you for the answer, because if you do it according to the documentation there is a confusion, and I need the chat room to receive just output by 443 without using port 80.
